### PR TITLE
Native Auth Backward Compatibility feature, Fixes AB#3287604

### PR DESCRIPTION
**Pull request by Yuki-YuXin**
Fixes [AB#{3287604}](https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/3287604)
Related common PR: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2669

Native Auth Backward Compatibility feature includes:
- [Add capabilities parameter to native auth public client application initialization, Fixes AB#3287412](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305/commits/6d553466cab96eb9642f5dd65e63d3492b95d8b3)
- [Handle redirect across all endpoints and return the redirect reason to the developers, Fixes AB#3288191](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305/commits/a121610861b63f5b7b7fe3c7a5240d2c0ee4c0ea)
-  [Use parameter instead of config when initialization](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2313)

The main addition to this PR is the testing section

### CODE REVIEW

**Comment by Yuki-YuXin** [comment]
A hint for the change log update and sub module reference after common PR get merged.

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java [hunk 1]
```diff
@@ -247,6 +248,9 @@ public void setClientId(final String clientId) {
      * @return The List of current Authorities.
      */
     public List<Authority> getAuthorities() {
+        if (mAuthorities == null) {
+            mAuthorities = new ArrayList<>();
+        }
         return mAuthorities;
```
Comment (Yuki-YuXin) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java]: May need Redmon team for confirmation since it's an important class. Or maybe there's another, better way to fix it.

**_Why the changes?_**
1. developer's authorities config cannot be null. 
2. the default status of authorities is null.
3. `getAuthorities()` throws error when authorities is null.
4. The transformation of null to empty list depends on the `defaultMsalConfiguration(R.raw.msal_default_config)` and merging, clearing operations.

_**What's the problem?**_
```kotlin
val nativeAuthConfig = NativeAuthPublicClientApplicationConfiguration()
            nativeAuthConfig.clientId = config.clientId
            val authorityObject = Authority.getAuthorityFromAuthorityUrl(
                config.authorityUrl,
                config.clientId
            )
            authorityObject.setDefault(true)
            nativeAuthConfig.getAuthorities().add(authorityObject)
            nativeAuthConfig.setChallengeTypes(challengeTypes)
            nativeAuthConfig.setCapabilities(capabilities)
```
`nativeAuthConfig.getAuthorities()` will have null pointer error if authorities is null. However, `Authorities` does not open the set method. 

Previous deprecated method
```java
return createNativeAuthApplication(
    Companion.initializeNativeAuthConfiguration(context),
    clientId,
    authority,
    redirectUri,
    challengeTypes
);
```
which use 
```java
fun initializeNativeAuthConfiguration(context: Context): NativeAuthPublicClientApplicationConfiguration {
    return initializeNativeAuthConfigurationInternal(context, null)
}
```
Since the `developerConfig` parameter is null, thus will skip the `developerConfig.validateConfiguration()` step inside `initializeNativeAuthConfigurationInternal(context: Context, developerConfig: NativeAuthPublicClientApplicationConfiguration?)` method, where `if (authorities == null || authorities.size == 0)` will throw an exception.

The authorities keep null until merging and clearing
```java
private fun initializeNativeAuthConfigurationInternal(context: Context, developerConfig: NativeAuthPublicClientApplicationConfiguration?): NativeAuthPublicClientApplicationConfiguration {
            val defaultMsalConfiguration: PublicClientApplicationConfiguration = PublicClientApplicationConfigurationFactory.loadConfiguration(context, R.raw.msal_default_config)
            // one authorities

            val defaultNativeAuthConfiguration = loadDefaultNativeAuthConfiguration(context)
           // no authorities

            val config = NativeAuthPublicClientApplicationConfiguration()
            config.appContext = context

            // Merge base MSAL defaults
            config.mergeConfiguration(defaultMsalConfiguration)  // authorities.size() == 1
            // Marge Native Auth defaults
            config.mergeConfiguration(defaultNativeAuthConfiguration)  // authorities.size() == 1

            config.authorities.clear()  // authorities.size() == 0 but authorities is not null

            // Check for developerConfig
            if (developerConfig != null) {  // skip the validation since developer config is null
                // Merge developer Native Auth configuration
                config.mergeConfiguration(developerConfig)
                // Validate Native Auth configuration
                config.validateConfiguration()
            }
```
Then structure the authority from the URL and append it into the authorities list
```kotlin
private static NativeAuthPublicClientApplication createNativeAuthApplication(@NonNull final NativeAuthPublicClientApplicationConfiguration config,
                                                                                 @Nullable final String clientId,
                                                                                 @Nullable final String authorityURL,
                                                                                 @Nullable final String redirectUri,
                                                                                 @Nullable final List<String> challengeTypes) throws BaseException {

        if (authorityURL != null) {
            config.getAuthorities().clear();

            final Authority authorityObject = Authority.getAuthorityFromAuthorityUrl(authorityURL, clientId);
            authorityObject.setDefault(true);
            config.getAuthorities().add(authorityObject);
        }
```

When passing in Config using ConfigFile and ResourceId, if `authoriries` field doesn't exist in the JSON file, the `gson.fromJson(config, NativeAuthPublicClientApplicationConfiguration::class.java)` will initialize it to null and fail the validation.

**Comment by Yuki-YuXin** [comment]
Fixes [AB#{3287604}](https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/3287604)

File: msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/SignInStates.kt [hunk 2]
```diff
@@ -287,7 +287,17 @@ class SignInCodeRequiredState internal constructor(
                         )
                     }
 
-                    is INativeAuthCommandResult.Redirect, is INativeAuthCommandResult.APIError -> {
+                    is INativeAuthCommandResult.Redirect -> {
```
[the remark this answers is not in the document]
Comment (Yuki-YuXin) [reply msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/SignInStates.kt:290]: The repetition is to enumerate the possible scenarios under each method. Tend to keep the existing.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfiguration.kt [hunk 3]
```diff
@@ -194,6 +218,28 @@ public class NativeAuthPublicClientApplicationConfiguration :
         }
     }
 
+    /**
+     * Validates that the capabilities passed are valid
+     */
+    @Throws(MsalClientException::class)
+    private fun validateCapabilities() {
+        // Make all capabilities lowercase for simplicity
+        capabilities = capabilities?.map { it.lowercase() }
```
[the remark this answers is not in the document]
Comment (Yuki-YuXin) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfiguration.kt:227]: VALID_CAPABILITIES are defined lowercase constants.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt [hunk 4]
```diff
@@ -331,4 +336,57 @@ class SignInMFATest : NativeAuthPublicClientApplicationAbstractTest() {
...
+            runBlocking {
+                val username = config.email
+                val password = getSafePassword()
+                val params = NativeAuthSignInParameters(username)
+                params.password = password.toCharArray()
+                params.claimsRequest = ClaimsRequest.getClaimsRequestFromJsonString(authenticationContextRequestClaimJson)
+
+                // SignIn specifying authentication context as claim
+                val result = application.signIn(params)
+                assertResult<SignInResult.MFARequired>(result)
```
Comment (nilo-ms) [inline msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:369]: the expected result here is "redirect" no "MFA_required"

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt [hunk 5]
```diff
@@ -331,4 +336,57 @@ class SignInMFATest : NativeAuthPublicClientApplicationAbstractTest() {
...
+    /**
+     * Full flow: Ensure that correct error is returned when an user doesn’t supply the “mfa_required” capability.
+     * - Initialise client with empty capabilities config
+     * - SignIn specifying authentication context as claim
+     * - Receive MFA required error from API
+     * - Request default challenge and submit correct challenge
+     * - Return redirect with reason mfa required was not supplied
+     *
+     */
+    @Ignore("Retrieving OTP code failure.")
```
Comment (nilo-ms) [inline msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt]: no OTP code should be sent for this test, so it should be enabled

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt [hunk 6]
```diff
@@ -290,4 +290,97 @@ class NativeAuthPublicClientApplicationConfigurationTest {
...
+        whenever(spyConfig.authorities).thenReturn(listOf(NativeAuthCIAMAuthority(ciamAuthority, clientId)))
+        whenever(spyConfig.defaultAuthority).thenReturn(NativeAuthCIAMAuthority(ciamAuthority, clientId))
+        whenever(spyConfig.isSharedDevice).thenReturn(false)
+        whenever(spyConfig.useBroker).thenReturn(false)
+        spyConfig.setCapabilities(listOf("mfa_required"))
+        spyConfig.validateConfiguration()
+    }
+
+    @Test
+    fun testRepeatedCorrectCapabilities() {
```
Comment (nilo-ms) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:376]: what this test is exactly testing? Where are you checking that capabilities duplicates are removed?

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java (hunk 1 above)
Comment (Yuki-YuXin) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java]: Please refer to https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2313 for the new solution.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt (hunk 5 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:349]: Done.

Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt]: Done. Use "Backward compatibility feature not available in eSTS production" instead.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt (hunk 4 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:369]: Done. And updated the others.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt (hunk 6 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:376]: That's a good point. Add `Assert.assertEquals(spyConfig.getCapabilities()?.size,2)` and also for challenge types.

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java [hunk 7]
```diff
@@ -82,6 +82,7 @@
 
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
```
Comment (mohitc1) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java]: Do you need this anymore?

**Comment by mohitc1** [comment]
Don't see any issues with the three files you asked for review. Let me know if there's was specific part you were looking review for. 

Can you get this in ASAP?

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java [hunk 8]
```diff
@@ -237,6 +237,7 @@ static class NONNULL_CONSTANTS {
         static final String AUTHORITY = "authority";
         static final String REDIRECT_URI = "redirect_uri";
         static final String CONFIG_FILE = "config_file";
+        static final String CLIENT_PARAMETER = "client_parameter";
```
Comment (fadidurah) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:240]: Client_parameter? Should this be "capabilities"?

**Review by fadidurah — COMMENTED** [review]
Reviewed MSAL-only code and LGTM, no review done for native auth code.

**Review by fadidurah — APPROVED** [review]

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java (hunk 7 above)
Comment (Yuki-YuXin) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java]: After checking. Nope. Removed this excess lines.

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java (hunk 8 above)
Comment (Yuki-YuXin) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java:240]: In this PR, a new parameter class is introduced to all those config parameters (also capabilities) to provide extension for the future.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt [hunk 9]
```diff
@@ -245,4 +247,49 @@ class SignInJITTest : NativeAuthPublicClientApplicationAbstractTest() {
...
+
+                // SignIn after signUp with authentication context as claims to trigger MFA
+                val continuationParameters = NativeAuthSignInContinuationParameters()
+                continuationParameters.claimsRequest = ClaimsRequest.getClaimsRequestFromJsonString(authenticationContextRequestClaimJson)
+                val signWithContinuationResult = (submitCodeResult as SignUpResult.Complete).nextState.signIn(continuationParameters)
+
+                // Return redirect with reason registration required was not supplied
+                assertTrue(signWithContinuationResult is SignInError)
+                assertTrue((signWithContinuationResult as SignInError).isBrowserRequired())
+                assertTrue(signWithContinuationResult.errorMessage!!.contains("registration required was not supplied"))
```
Comment (spetrescu84) [inline msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt:291]: Is this the message that comes back from the server? The format looks strange.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt [hunk 10]
```diff
@@ -331,4 +336,43 @@ class SignInMFATest : NativeAuthPublicClientApplicationAbstractTest() {
...
+                params.claimsRequest = ClaimsRequest.getClaimsRequestFromJsonString(authenticationContextRequestClaimJson)
+
+                // SignIn specifying authentication context as claim
+                val result = application.signIn(params)
+                assertResult<SignInResult.MFARequired>(result)
+
+                // Return redirect with reason mfa required was not supplied
+                assertTrue(result is SignInError)
+                assertTrue((result as SignInError).isBrowserRequired())
+                assertTrue(result.errorMessage!!.contains("mfa required was not supplied"))
```
Comment (spetrescu84) [inline msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:374]: Is this the message that comes back from the server? The format looks strange.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt [hunk 11]
```diff
@@ -289,5 +290,100 @@ class NativeAuthPublicClientApplicationConfigurationTest {
...
+    fun testInvalidCapabilities() {
+        val config = NativeAuthPublicClientApplicationConfiguration()
+        config.clientId = clientId
+        config.accountMode = AccountMode.SINGLE
+        val spyConfig = spy(config)
+        whenever(spyConfig.authorities).thenReturn(listOf(NativeAuthCIAMAuthority(ciamAuthority, clientId)))
+        whenever(spyConfig.defaultAuthority).thenReturn(NativeAuthCIAMAuthority(ciamAuthority, clientId))
+        whenever(spyConfig.isSharedDevice).thenReturn(false)
+        whenever(spyConfig.useBroker).thenReturn(false)
+        spyConfig.setCapabilities(listOf("lorem"))
```
Comment (spetrescu84) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:321]: Is this part of another PR where only specific capabilities work?

Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:321]: Pardon. This is part of the sub-pr under https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2305/commits/6d553466cab96eb9642f5dd65e63d3492b95d8b3#diff-1e72e28f17fc1763d76a4b809b06adc0afb3fbf660cde3f7374282ab0547359c

Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:321]: "lorem" is kind of invalid string practice I referred from other test cases.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt (hunk 9 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt:291]: This is a placeholder when the test slice didn't work fine before. 
Done. Updated with "missing registration_required" which is more generic.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt (hunk 10 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:374]: This is a placeholder when the test slice didn't work fine before. 
Done. Updated with "missing registration_required" which is more generic.

Comment (spetrescu84) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:374]: So the test slice returns "missing registration_required" or we can't test? I see @nilo-ms [only checks that is not nil](https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/2648/files). Maybe do that also here? I am saying this that once we remove the ignore above this test will automatically start working. We can leave it to fix it then.

Comment (spetrescu84) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:374]: Also just in case, this one is about MFA.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt (hunk 9 above)
Comment (spetrescu84) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt:291]: So the test slice returns "missing registration_required" or we can't test? I see @nilo-ms [only checks that is not nil](https://github.com/AzureAD/microsoft-authentication-library-for-objc/pull/2648/files). Maybe do that also here? I am saying this that once we remove the ignore above this test will automatically start working. We can leave it to fix it then.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt (hunk 11 above)
Comment (spetrescu84) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationConfigurationTest.kt:321]: Found this check, all good - `if (capability !in VALID_CAPABILITIES) {`

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt (hunk 9 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInJITTest.kt:291]: Screenshot 2025-06-23 112641
Checking is not nil is a vague choice after the server response is specific.

File: msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt (hunk 10 above)
Comment (Yuki-YuXin) [reply msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt:374]: Yep. missing mfa_required

**Review by spetrescu84 — APPROVED** [review]
